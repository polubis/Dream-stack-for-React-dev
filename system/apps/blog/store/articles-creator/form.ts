import {
  NullableValidatorType,
  form,
  maxLength,
  minLength,
  pascalCase,
  required,
} from '@system/utils';
import type { ArticlesCreatorStore } from './defs';
import type { ArticleTag } from '@system/blog-api-models';

const noDuplicatedTags = (
  tag: ArticleTag,
  values: ArticlesCreatorStore.TagFormData
): NullableValidatorType => {
  return values.tags.map((tag) => tag.toLowerCase()).includes(tag.toLowerCase())
    ? 'duplicated'
    : null;
};

const tagForm = form<ArticlesCreatorStore.TagFormData>({
  validateOnInit: true,
})({
  tag: [required, minLength(1), maxLength(20), pascalCase, noDuplicatedTags],
});

const creatorForm = form<ArticlesCreatorStore.FormData>({
  validateOnInit: true,
})({
  title: [required],
  description: [required],
  content: [required],
  lang: [required],
  tags: [minLength(1), maxLength(10)],
});

export { creatorForm, tagForm };
