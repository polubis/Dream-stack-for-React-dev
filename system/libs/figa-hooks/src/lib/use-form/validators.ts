const run = (invalid: boolean, message: string): string =>
  invalid ? message : '';

export const required =
  (message = (value: string) => 'This field is required') =>
  (value: string) =>
    run(value === '', message(value));

export const minLength =
  (
    limit: number,
    message = (limit: number, value: string) => `Minimum length is ${limit}`
  ) =>
  (value: string) =>
    run(value.length > 0 && value.length < limit, message(limit, value));

export const maxLength =
  (
    limit: number,
    message = (limit: number, value: string) => `Maximum length is ${limit}`
  ) =>
  (value: string) =>
    run(value.length > 0 && value.length > limit, message(limit, value));

export const email =
  (message = () => 'Invalid email format') =>
  (value: string) =>
    run(
      !/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(
        value.toLowerCase()
      ),
      message()
    );
