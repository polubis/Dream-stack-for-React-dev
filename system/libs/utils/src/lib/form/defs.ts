type ValidatorType =
  | 'required'
  | 'minLength'
  | 'maxLength'
  | 'min'
  | 'max'
  | 'email'
  | 'pascalCase'
  | 'duplicated';

type NullableValidatorType = ValidatorType | null;

type FormTouch<Values extends FormValues> = {
  [K in keyof Values]: boolean;
};

interface FormMetadata<Values extends FormValues> {
  touched: boolean;
  untouched: boolean;
  dirty: boolean;
  pristine: boolean;
  touch: FormTouch<Values>;
}

type FormResult<Values extends FormValues> = {
  [K in keyof Values]: boolean;
};

type FormErrors<Values extends FormValues> = {
  [K in keyof Values]: NullableValidatorType;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type FormValues = Record<string | number | symbol, any>;

type FormState<Values extends FormValues> = {
  values: Values;
  errors: FormErrors<Values>;
  result: FormResult<Values>;
  invalidCount: number;
  validCount: number;
  invalid: boolean;
  valid: boolean;
  progress: number;
  keys: (keyof Values)[];
} & FormMetadata<Values>;

type FormValidatorsSetup<Values extends FormValues> = {
  [K in keyof Values]?: ((
    value: Values[K],
    state: Values
  ) => NullableValidatorType)[];
};

interface FormConfig {
  validateOnInit?: boolean;
}

export type {
  FormValidatorsSetup,
  FormState,
  FormErrors,
  FormResult,
  FormMetadata,
  FormTouch,
  FormConfig,
  NullableValidatorType,
  ValidatorType,
  FormValues,
};
