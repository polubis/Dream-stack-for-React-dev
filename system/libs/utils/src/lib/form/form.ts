import type {
  FormValidatorsSetup,
  FormErrors,
  FormMetadata,
  FormResult,
  FormState,
  FormTouch,
  FormValues,
  FormConfig,
} from './defs';

const createFormState = <
  Values extends FormValues,
  Validators extends FormValidatorsSetup<Values>
>(
  currentValues: Values,
  newValues: Values,
  validators: Validators,
  meta: FormMetadata<Values>
): FormState<Values> => {
  let invalid = false;
  const errors = {} as FormErrors<Values>;
  const result = {} as FormResult<Values>;
  let invalidCount = 0;
  const keys = Object.keys(newValues) as (keyof Values)[];

  for (const key in newValues) {
    const fns = validators[key] ?? [];

    if (fns.length > 0) {
      for (let i = 0; i < fns.length; i++) {
        const validator = fns[i];
        errors[key] = validator(newValues[key], currentValues);
        result[key] = !!errors[key];

        if (result[key]) {
          invalid = true;
          invalidCount++;
          break;
        }
      }
    } else {
      errors[key] = null;
      result[key] = false;
    }
  }

  const validCount = keys.length - invalidCount;

  return {
    errors,
    values: newValues,
    invalid,
    keys,
    valid: !invalid,
    result,
    invalidCount,
    validCount,
    progress: (validCount / keys.length) * 100,
    ...meta,
  };
};

const init = <Values extends FormValues>(values: Values): FormState<Values> => {
  const keys = Object.keys(values) as (keyof Values)[];

  return {
    values,
    touched: false,
    untouched: true,
    pristine: true,
    dirty: false,
    keys,
    result: keys.reduce<FormResult<Values>>((acc, key) => {
      acc[key] = false;
      return acc;
    }, {} as FormResult<Values>),
    errors: keys.reduce<FormErrors<Values>>((acc, key) => {
      acc[key] = null;
      return acc;
    }, {} as FormErrors<Values>),
    touch: keys.reduce<FormTouch<Values>>((acc, key) => {
      acc[key] = false;
      return acc;
    }, {} as FormTouch<Values>),
    progress: 0,
    valid: true,
    invalid: false,
    invalidCount: 0,
    validCount: keys.length,
  };
};

const form =
  <Values extends FormValues>(config?: FormConfig) =>
  <Validators extends FormValidatorsSetup<Values>>(validators: Validators) => {
    const set =
      (state: FormState<Values>) =>
      (newValues: Partial<Values>): FormState<Values> => {
        const keys = Object.keys(newValues) as Partial<keyof Values>[];

        return createFormState<Values, Validators>(
          state.values,
          {
            ...state.values,
            ...newValues,
          },
          validators,
          {
            pristine: state.pristine,
            touched: true,
            untouched: false,
            dirty: state.dirty,
            touch: {
              ...state.touch,
              ...keys.reduce<Partial<FormTouch<Values>>>((acc, key) => {
                acc[key] = true;
                return acc;
              }, {}),
            },
          }
        );
      };
    return {
      init: (values: Values) =>
        config?.validateOnInit ? set(init(values))({}) : init(values),
      set,
      confirm: (state: FormState<Values>): FormState<Values> => {
        return createFormState<Values, Validators>(
          state.values,
          state.values,
          validators,
          {
            dirty: true,
            pristine: false,
            touched: state.touched,
            untouched: state.untouched,
            touch: { ...state.touch },
          }
        );
      },
      toFormData: (state: FormState<Values>): FormData => {
        const formData = new FormData();

        Object.entries(state.values).forEach(([key, value]) => {
          formData.append(key, value);
        });

        return formData;
      },
    };
  };

export { form };
