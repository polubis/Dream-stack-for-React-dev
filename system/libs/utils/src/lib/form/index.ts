export * from './form';
export * from './defs';
