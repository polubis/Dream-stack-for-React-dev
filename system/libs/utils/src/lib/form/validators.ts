import type { NullableValidatorType } from './defs';

const required = (value: string | null | undefined): NullableValidatorType => {
  return !value ? 'required' : null;
};

const min =
  (limit: number) =>
  (value: number): NullableValidatorType => {
    return value < limit ? 'min' : null;
  };

const max =
  (limit: number) =>
  (value: number): NullableValidatorType => {
    return value > limit ? 'max' : null;
  };

const minLength =
  (limit: number) =>
  (value: string | unknown[]): NullableValidatorType => {
    return value.length < limit ? 'minLength' : null;
  };

const maxLength =
  (limit: number) =>
  (value: string | unknown[]): NullableValidatorType => {
    return value.length > limit ? 'maxLength' : null;
  };

export { required, minLength, maxLength, min, max };
