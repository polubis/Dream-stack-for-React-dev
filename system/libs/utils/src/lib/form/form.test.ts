import { form } from './form';
import { max, min, minLength, maxLength, required } from './validators';

describe('Form can be used when: ', () => {
  interface ExampleForm {
    username: string;
    password: string;
    code: number;
    flags: boolean[];
  }

  it('initializes state', () => {
    const { init } = form<ExampleForm>()({
      username: [required],
      password: [minLength(2)],
      code: [max(2)],
      flags: [],
    });

    expect(
      init({
        username: '',
        password: '',
        code: 1,
        flags: [true, false],
      })
    ).toMatchSnapshot();
  });

  it('allows to change field value and validates', () => {
    const { init, set } = form<ExampleForm>()({
      username: [required],
      password: [minLength(2)],
      code: [max(2)],
      flags: [],
    });

    const state = init({
      username: '',
      password: '',
      code: 3,
      flags: [true, false],
    });

    expect(
      set(state)({
        username: '',
        password: 'd',
        code: 4,
      })
    ).toMatchSnapshot();
  });

  it('validates on init if passed in configuration', () => {
    const { init } = form<ExampleForm>({ validateOnInit: true })({
      username: [required],
      password: [minLength(2)],
      code: [max(2)],
      flags: [],
    });

    expect(
      init({
        username: '',
        password: '',
        code: 3,
        flags: [true, false],
      })
    ).toMatchSnapshot();
  });

  it('allows to confirm and validates', () => {
    const { init, confirm } = form<ExampleForm>()({
      username: [required],
      password: [minLength(2)],
      code: [max(2)],
      flags: [],
    });

    const state = init({
      username: '',
      password: '',
      code: 3,
      flags: [true, false],
    });

    expect(confirm(state)).toMatchSnapshot();
  });

  it('converts values to form data format', () => {
    const { init, toFormData } = form<ExampleForm>()({
      username: [required],
      password: [minLength(2)],
      code: [max(2)],
      flags: [],
    });

    const state = init({
      username: 'username',
      password: '',
      code: 3,
      flags: [true, false],
    });

    expect(toFormData(state).getAll('username')).toEqual(['username']);
    expect(toFormData(state).get('username')).toEqual('username');
  });
});
