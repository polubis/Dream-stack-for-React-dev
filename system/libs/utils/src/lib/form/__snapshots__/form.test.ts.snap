// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form can be used when:  allows to change field value and validates 1`] = `
Object {
  "dirty": false,
  "errors": Object {
    "code": "max",
    "flags": null,
    "password": "minLength",
    "username": "required",
  },
  "invalid": true,
  "invalidCount": 3,
  "keys": Array [
    "username",
    "password",
    "code",
    "flags",
  ],
  "pristine": true,
  "progress": 25,
  "result": Object {
    "code": true,
    "flags": false,
    "password": true,
    "username": true,
  },
  "touch": Object {
    "code": true,
    "flags": false,
    "password": true,
    "username": true,
  },
  "touched": true,
  "untouched": false,
  "valid": false,
  "validCount": 1,
  "values": Object {
    "code": 4,
    "flags": Array [
      true,
      false,
    ],
    "password": "d",
    "username": "",
  },
}
`;

exports[`Form can be used when:  allows to confirm and validates 1`] = `
Object {
  "dirty": true,
  "errors": Object {
    "code": "max",
    "flags": null,
    "password": "minLength",
    "username": "required",
  },
  "invalid": true,
  "invalidCount": 3,
  "keys": Array [
    "username",
    "password",
    "code",
    "flags",
  ],
  "pristine": false,
  "progress": 25,
  "result": Object {
    "code": true,
    "flags": false,
    "password": true,
    "username": true,
  },
  "touch": Object {
    "code": false,
    "flags": false,
    "password": false,
    "username": false,
  },
  "touched": false,
  "untouched": true,
  "valid": false,
  "validCount": 1,
  "values": Object {
    "code": 3,
    "flags": Array [
      true,
      false,
    ],
    "password": "",
    "username": "",
  },
}
`;

exports[`Form can be used when:  initializes state 1`] = `
Object {
  "dirty": false,
  "errors": Object {
    "code": null,
    "flags": null,
    "password": null,
    "username": null,
  },
  "invalid": false,
  "invalidCount": 0,
  "keys": Array [
    "username",
    "password",
    "code",
    "flags",
  ],
  "pristine": true,
  "progress": 0,
  "result": Object {
    "code": false,
    "flags": false,
    "password": false,
    "username": false,
  },
  "touch": Object {
    "code": false,
    "flags": false,
    "password": false,
    "username": false,
  },
  "touched": false,
  "untouched": true,
  "valid": true,
  "validCount": 4,
  "values": Object {
    "code": 1,
    "flags": Array [
      true,
      false,
    ],
    "password": "",
    "username": "",
  },
}
`;

exports[`Form can be used when:  validates on init if passed in configuration 1`] = `
Object {
  "dirty": false,
  "errors": Object {
    "code": "max",
    "flags": null,
    "password": "minLength",
    "username": "required",
  },
  "invalid": true,
  "invalidCount": 3,
  "keys": Array [
    "username",
    "password",
    "code",
    "flags",
  ],
  "pristine": true,
  "progress": 25,
  "result": Object {
    "code": true,
    "flags": false,
    "password": true,
    "username": true,
  },
  "touch": Object {
    "code": false,
    "flags": false,
    "password": false,
    "username": false,
  },
  "touched": true,
  "untouched": false,
  "valid": false,
  "validCount": 1,
  "values": Object {
    "code": 3,
    "flags": Array [
      true,
      false,
    ],
    "password": "",
    "username": "",
  },
}
`;
